linera-core/src/client/mod.rs

```diff
@@ -825,149 +831,164 @@ impl<Env: Environment> Client<Env> {
-                    other_sender_chains.push(sender_chain_id);
-                    return None;
-                };
-                Some(async move {
-                    let certificates = remote_node
-                        .download_certificates_by_heights(sender_chain_id, remote_heights)
-                        .await?;
-                    Ok::<Vec<_>, ChainClientError>(certificates)
-                })
-            },
-        ))
-        .buffer_unordered(self.options.max_joined_tasks)
-        .try_collect::<Vec<_>>()
-        .await?
-        .into_iter()
-        .flatten()
-        .collect::<Vec<_>>();
-
-        let mut certificates_by_height_by_chain = BTreeMap::new();
-
-        // Check the signatures and keep only the ones that are valid.
-        for confirmed_block_certificate in certificates {
-            let block_header = &confirmed_block_certificate.inner().block().header;
-            let sender_chain_id = block_header.chain_id;
-            let height = block_header.height;
-            let epoch = block_header.epoch;
-            match Self::check_certificate(max_epoch, &committees, &confirmed_block_certificate)? {
-                CheckCertificateResult::FutureEpoch => {
-                    warn!(
-                        "Postponing received certificate from {sender_chain_id:.8} at height \
-                         {height} from future epoch {epoch}"
-                    );
-                    // Do not process this certificate now. It can still be
-                    // downloaded later, once our committee is updated.
-                }
-                CheckCertificateResult::OldEpoch => {
-                    // This epoch is not recognized any more. Let's skip the certificate.
-                    // If a higher block with a recognized epoch comes up later from the
-                    // same chain, the call to `receive_certificate` below will download
-                    // the skipped certificate again.
-                    warn!("Skipping received certificate from past epoch {epoch:?}");
-                }
-                CheckCertificateResult::New => {
-                    certificates_by_height_by_chain
-                        .entry(sender_chain_id)
-                        .or_insert_with(BTreeMap::new)
-                        .insert(height, confirmed_block_certificate);
+            for certificate in certificates {
+                let hash = certificate.hash();
+                let chain_id = certificate.block().header.chain_id;
+                let height = certificate.block().header.height;
+                // We checked the certificates right after downloading them.
+                let mode = ReceiveCertificateMode::AlreadyChecked;
+                if let Err(error) = self
+                    .receive_sender_certificate(certificate, mode, None)
+                    .await
+                {
+                    error!(%error, %hash, "Received invalid certificate");
```

Comment:
Could a faulty validator trigger this by refusing to supply a block published by that certificate? (See the scenario above: It's not guaranteed that any honest validator has this certificate.)